### PR TITLE
fix: create upload path from env

### PR DESCRIPTION
**Pull request by ghost**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](../../CONTRIBUTING.md#submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [x] I have performed preliminary unit testing.
- [ ] I have updated the required [api docs](../../apps/api/docs/) as applicable.
- [x] I have added/updated test cases to the [test suite](../../apps/api/TranscriptSummarizer.postman_collection.json) as applicable

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [ ] Related changes have been added (optional)
- [ ] Screenshots have been added (optional)
- [ ] Query request and response examples have been added (as applicable, in case added or updated)
- [ ] Documentation changes have been listed (as applicable)
- [ ] Test suite output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

---

**Description:**

Added a fix for upload folder not existing issue and postman collection json is updated with response.
**Related changes:**

- Add short points about the different changes made within the files in this PR.

**Screenshots:**

Add any screenshots as required.

**Query request and response:**

- Add any query request body, cURL statement and response body for the made change or addition.

**Documentation changes:**

- Add a list of changes made to the API documents with brief descriptions.

**Test suite output:**

- Add the output of the status of different test cases in the testing suite.

**Pending actions:**

- Add list of any pending actions that have or would require to be done in this PR.

**Additional notes:**

- Add list of any additional notes you may want to convey in this PR.

## Summary by CodeRabbit

- **New Features**
  - Added configuration option `Upload_Path` for specifying the upload folder location.

- **Refactor**
  - Enhanced file system operations in `SummaryService` using new libraries and dynamic path configuration.

### CODE REVIEW

File: apps/api/.env.example [hunk 1]
```diff
@@ -4,6 +4,9 @@ SERVER_PORT=
 # Node env
 NODE_ENV= # Use "development" for graphql playground to work
 
+# Upload folder location
+Upload_Path=./src/uploads/
```
Comment (xixas) [inline apps/api/.env.example:8]: This is incorrect. We need to provide full path here. Not relative path.
Also that variable is inconsistent with others

File: apps/api/src/modules/summary/summary.service.ts [hunk 2]
```diff
@@ -27,9 +33,11 @@ export class SummaryService extends CoreService<Summary> {
     if (inputFile) {
       const { createReadStream, filename } = await inputFile;
 
-      const uniqueIdentifier = uuidv4();
+      const uniqueIdentifier = uuidv4().replace(/-/g, '').substring(0, 12);
       const modifiedFilename = `${uniqueIdentifier}_${filename}`;
-      const fileLocation = join(process.cwd(), `./src/upload/${modifiedFilename}`);
+      const fileLocation = join(process.cwd(), configService.getOrThrow('Upload_Path'), modifiedFilename);
```
Comment (xixas) [inline apps/api/src/modules/summary/summary.service.ts:38]: We won't use cwd.
We directly use the path from env value.
You can use process.cwd() + /upload as default folder though.

File: apps/api/.env.example [hunk 3]
```diff
@@ -4,6 +4,9 @@ SERVER_PORT=
...
+
 # Database configuration
 DB_TYPE=
 DB_HOST=
```
Comment (xixas) [inline apps/api/.env.example:12]: @ramraut-osm 
Can you add this change in your PR instead. And fix the prettier lintings.
https://github.com/OsmosysSoftware/osm-transcript-summarizer/pull/17#discussion_r1597449894

**Review by xixas — CHANGES_REQUESTED** [review]

**Comment by xixas** [comment]
@ramraut-osm  Title doesn't make sense, it says feat but says fix for upload path issue.
Also "for upload path issue" => This doesn't give any information.
Just make it
fix: create upload path from env.

**Comment by ramraut-osm** [comment]
> @ramraut-osm Title doesn't make sense, it says feat but says fix for upload path issue. Also "for upload path issue" => This doesn't give any information. Just make it fix: create upload path from env.

Updated

File: apps/api/.env.example (hunk 3 above)
Comment (ramraut-osm) [reply apps/api/.env.example:12]: Done

File: apps/api/.env.example (hunk 1 above)
Comment (ramraut-osm) [reply apps/api/.env.example:8]: Updated

File: apps/api/src/modules/summary/summary.service.ts (hunk 2 above)
Comment (ramraut-osm) [reply apps/api/src/modules/summary/summary.service.ts:38]: Added the changes, removed process.cwd()

File: apps/api/.env.example (hunk 3 above)
Comment (xixas) [reply apps/api/.env.example:12]: Env naming is still incorrect.

File: apps/api/src/modules/summary/summary.service.ts [hunk 4]
```diff
@@ -24,19 +25,25 @@ export class SummaryService extends CoreService<Summary> {
...
+      await fs.ensureDir(process.env.Upload_Path);
 
       return new Promise((resolve, reject) => {
         createReadStream()
           .pipe(createWriteStream(fileLocation))
           .on('finish', async () => {
             const summary = this.summaryRepository.create({ inputFile: modifiedFilename });
+
             try {
+              this.logger.log('Saving Uploaded File Name...');
```
Comment (xixas) [inline apps/api/src/modules/summary/summary.service.ts:46]: Should be  this.logger.log('Saving Uploaded File details...');

**Review by xixas — CHANGES_REQUESTED** [review]

File: apps/api/src/modules/summary/summary.service.ts (hunk 2 above)
Comment (xixas) [reply apps/api/src/modules/summary/summary.service.ts]: Please do all the changes like asked above.

You did not do
- You can use process.cwd() + /upload as default folder though.

File: apps/api/.env.example (hunk 3 above)
Comment (ramraut-osm) [reply apps/api/.env.example:12]: Changed it to UPLOAD_FOLDER_PATH

File: apps/api/src/modules/summary/summary.service.ts (hunk 4 above)
Comment (ramraut-osm) [reply apps/api/src/modules/summary/summary.service.ts:46]: Updated

File: apps/api/src/modules/summary/summary.service.ts (hunk 2 above)
Comment (ramraut-osm) [reply apps/api/src/modules/summary/summary.service.ts]: Added process.cwd() + /upload as fallback condition

File: apps/api/src/modules/summary/summary.service.ts [hunk 5]
```diff
@@ -24,19 +25,27 @@ export class SummaryService extends CoreService<Summary> {
...
 
-      const uniqueIdentifier = uuidv4();
+      const uniqueIdentifier = uuidv4().replace(/-/g, '').substring(0, 10);
       const modifiedFilename = `${uniqueIdentifier}_${filename}`;
-      const fileLocation = join(process.cwd(), `./src/upload/${modifiedFilename}`);
+      const fileLocation = process.env.Upload_Path
+        ? join(process.env.Upload_Path, modifiedFilename)
+        : join(process.cwd(), `./src/uploads/${modifiedFilename}`);
+
+      await fs.ensureDir(process.env.Upload_Path || join(process.cwd(), './src/uploads/'));
```
[the remark this answers is not in the document]
Comment (ramraut-osm) [reply apps/api/src/modules/summary/summary.service.ts]: Validated and sanitized the upload folder path in latest commit

File: apps/api/src/modules/summary/summary.service.ts [hunk 6]
```diff
@@ -31,9 +32,24 @@ export class SummaryService extends CoreService<Summary> {
     if (inputFile) {
       const { createReadStream, filename } = await inputFile;
 
-      const uniqueIdentifier = uuidv4();
+      const uniqueIdentifier = uuidv4().replace(/-/g, '').substring(0, 10);
       const modifiedFilename = `${uniqueIdentifier}_${filename}`;
-      const fileLocation = join(process.cwd(), `./src/upload/${modifiedFilename}`);
+      const uploadPath = process.env.Upload_Path
```
Comment (xixas) [inline apps/api/src/modules/summary/summary.service.ts:37]: Fix this, and we use configService not process.env.
defaultPath = join(process.cwd(), 'src', 'uploads');
Don't keep it inside src folder it should be in root along with logs process.cwd() + /uploads like I mentioned.

File: apps/api/src/modules/summary/summary.service.ts [hunk 7]
```diff
@@ -31,9 +32,24 @@ export class SummaryService extends CoreService<Summary> {
...
+        ? process.env.Upload_Path.replace(/[^\w\s/]/g, '')
+        : null;
+
+      if (uploadPath) {
+        const absoluteUploadPath = resolve(uploadPath);
+        await fs.ensureDir(absoluteUploadPath);
+      } else {
+        const defaultPath = join(process.cwd(), 'src', 'uploads');
```
Comment (xixas) [inline apps/api/src/modules/summary/summary.service.ts:45]: Keep the path value as a constant and reuse it.
process.cwd(), 'src', 'uploads'

File: apps/api/src/modules/summary/summary.service.ts [hunk 8]
```diff
@@ -31,9 +32,24 @@ export class SummaryService extends CoreService<Summary> {
...
+        await fs.ensureDir(defaultPath);
+      }
+
+      const fileLocation = join(
+        uploadPath || resolve(process.cwd(), 'src', 'uploads'),
```
Comment (xixas) [inline apps/api/src/modules/summary/summary.service.ts:50]: Keep the path value as a constant and reuse it.
We should just have one upload path.
Also make sure you are creating the folder in  main.ts along with log folder first.
Its good to ensureDir as a rare case when someone deletes the folder manually.

**Review by xixas — CHANGES_REQUESTED** [review]

File: apps/api/src/modules/summary/summary.service.ts (hunk 6 above)
Comment (ramraut-osm) [reply apps/api/src/modules/summary/summary.service.ts:37]: Updated to use configservice

File: apps/api/src/modules/summary/summary.service.ts (hunk 7 above)
Comment (ramraut-osm) [reply apps/api/src/modules/summary/summary.service.ts:45]: Done

File: apps/api/src/modules/summary/summary.service.ts (hunk 8 above)
Comment (ramraut-osm) [reply apps/api/src/modules/summary/summary.service.ts:50]: Done

File: apps/api/src/main.ts [hunk 9]
```diff
@@ -8,11 +8,16 @@ import { HttpExceptionFilter } from './common/http-exception.filter';
 import { JsendFormatter } from './common/jsend-formatter';
 
 const logDir = 'logs';
+const uploadDir = 'uploads';
```
Comment (xixas) [inline apps/api/src/main.ts:11]: This  is incorrect. What if the env value is added?
You are supposed to get the upload folder path and create the folder there.
Here you are only doing it for default path.

**Review by xixas — CHANGES_REQUESTED** [review]

File: apps/api/src/main.ts (hunk 9 above)
Comment (ramraut-osm) [reply apps/api/src/main.ts]: Updated the condition to use env value for folder path. Also tested for conditions when env value is missing.

**Review by xixas — APPROVED** [review]

**Pull request merged**